src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -824,13 +843,14 @@ static UInt64 getLimitUIntValue(const ASTPtr & node, ContextPtr context, const s
         throw Exception(
             "Illegal type " + type->getName() + " of " + expr + " expression, must be numeric type", ErrorCodes::INVALID_LIMIT_EXPRESSION);
 
-    Field converted = convertFieldToType(field, DataTypeUInt64());
+    Field converted = convertFieldToType(field, DataTypeInt128());
     if (converted.isNull())
         throw Exception(
-            "The value " + applyVisitor(FieldVisitorToString(), field) + " of " + expr + " expression is not representable as UInt64",
+            "The value " + applyVisitor(FieldVisitorToString(), field) + " of " + expr + " expression is not representable as Int128",
             ErrorCodes::INVALID_LIMIT_EXPRESSION);
+    Int128 res = converted.safeGet<Int128>();
 
-    return converted.safeGet<UInt64>();
+    return res >= 0 ? UInt64(res) : UInt64(-res);
```

Comment:
Converting `Int128 -> UInt64` can lead to some special effects:
```sql
:) select * from system.numbers LIMIT 18446744073709552616

SELECT *
FROM system.numbers
LIMIT 18446744073709552000.

Query id: f2d33e9d-dcbe-4464-92fc-171144d8817a

Ok.

0 rows in set. Elapsed: 0.008 sec.
```
We should check that `|res| < UINT64_MAX`